### ISSUE TITLE
How to pin a request to a connection/transaction?

### ISSUE THREAD

**Issue by jakepearson** [knex/knex#1681]
Hi,
I would like to be able to start a `transaction` at beginning of a request in a web server middleware and then `commit` or `rollback` the transaction when the request is completing in another middleware. I see that there is a `transaction` function on the `knex` object, but it requires passing it a code block execute so `knex` can handle the transaction lifecycle.

Is there a function to call to get an instance of a transaction that I can pass through my app and `commit` or `rollback` at the end?

Alternatively, we would like to make sure the whole request is on the same `connection`. Instead, is there a function to call to `acquire` a `connection`?

**Comment by tgriesser** [comment]
@jakepearson so this is the main thing i've been working on over here: https://github.com/tgriesser/knex/pull/1659

I was originally considering rolling a ton of changes into a 1.0 but I think the transaction / context stuff is decently solid and so I'm considering releasing it sometime soon along with #1647 as a 0.14-alpha or something along those lines while I iron out the edge cases / testing / semantics of some of the new additions. It only works with mysql / pg / sqlite3 for now, so I'm looking to see if I can fix/test the other clients before shipping.

What it does is add a new approach to dealing with transactions:

```js
// doesn't actually execute BEGIN until the first query, if the transaction is committed
// or rolled back without any queries, none are executed. This makes transactions 
// inexpensive to create for each request
const trx = knex.transaction() 

trx.select('*').from('accounts').where('id', 1).then(...)

trx.table('accounts').insert().then(...)

trx.commit() // or trx.rollback()
```

```js
// A context is a knex instance where all queries make a request on the same connection:
// Contexts & Transactions both also maintain a query log which can be 
// retrieved with `.getExecutionLog()`

const ctx = knex.context()

ctx.select('*').from('accounts').where('id', 1).then(...)

ctx.table('accounts').insert().then(...)

ctx.end()
```

What this then allows for is something like what you're after (probably needs to be setup a bit nicer):

```js
app.use((req, res, next) => {
  if (req.method === 'POST' || req.method === 'PUT' || req.method === 'DELETE') {
    const trx = req.transaction()
    req.knex = trx
    req.on('end', () => {
      if (!trx.isTransactionComplete()) {
        throw new Error('Must call .commit or / .rollback before request end')  
      }
    })
  } else {
    req.knex = req.context()
    req.on('end', () => {
      req.knex.end()
    })
  }
  next()
})
```

On the top level knex it adds the following:

```js
knex.context() // detailed above
knex.transaction() // new form detailed above
knex.isTransaction() // boolean, whether the current instance is a transaction
knex.isInTransaction() // boolean, whether we're in a transaction (for sub-contexts)
knex.isTransactionComplete() // boolean, whether the current trx client is finished
knex.isRootContext() // the root context is the top level knex
knex.getConnection() // returns promise w/ current connection for the context / transaction
knex.getExecutionLog() // returns info about queries executed for context / transaction
```

Interested to get some feedback to see if this is what you're after / if you'd be interested in kicking the tires on the "refactor" branch to see if there's stuff that's either confusing or needs work.

**Comment by jakepearson** [comment]
@tgriesser Thanks for getting back to me. This sounds just like what my team would like to do with transactions or connections. I think I am nervous about putting the feature branch into production, but I would be delighted to look through any PRs or documentation you have. I think I would be comfortable with putting the new code into my service once you have it in your master branch. Please let me what I can do to help, Jake

**Comment by tgriesser** [comment]
@jakepearson cool, I'll keep you posted.

**Comment by srmelody** [comment]
hey @tgriesser - huge fan of knex and a co-worker of @jakepearson 's.  We would be interested in taking a semi-baked 0.14-alpha with this and do some testing with it.

**Comment by wolfgang42** [comment]
I'd also like to beta-test this feature.  Is there a branch anywhere with this code at the moment?

**Comment by cjnqt** [comment]
Same here, any chance of doing that @tgriesser ?

**Comment by cjnqt** [comment]
> the transaction / context stuff is decently solid and so I'm considering releasing it sometime soon along with #1647 as a 0.14-alpha or something along those lines.

Can we use the refactor-branch for testing this? @tgriesser Super-excited about the new transaction features.

**Comment by taybin** [comment]
Any progress on this?  I'm trying to have knex wrap my integration tests in a transaction that is rolled back after each test, and I believe I need this functionality to make that possible.

**Comment by moll** [comment]
What I've done for the past years is limit the pool to 1 connection in tests and in a share before-each block call `knex.raw("BEGIN")` and later `knex.raw("ROLLBACK")`. Done this way, you can selectively use transactions in tests where applicable and switch to a classic `TRUNCATE` (for Pg) for those tests that make use of transactions themselves.

**Comment by taybin** [comment]
@moll I can't use `TRUNCATE` because my database has test data in it and I want to preserve that data.  Since this is Pg, and it doesn't support nested transactions, I'm feeling a little stuck or will have to reimplement the proposed 0.14-alpha transaction function mentioned a couple years ago.

**Comment by moll** [comment]
Well, I consider it a good practice to create test data fresh in every test, so TRUNCATE in a before-each for those non-transaction cases should be fine. If you're talking about seed-data, then you can probably reseed the database, again, for a known good before each block, couldn't you? You wouldn't want one test that went haywire from affecting any other test, after all.

Nested transactions in Pg take the form of SAVEPOINTs, but if you're doing transactions in the code you're testing, you'd indeed require query rewriting. I thought of that for my own use cases (I was using transactions for performance reasons), but couldn't be bothered as TRUNCATE was fast enough and fortunately I don't need seed data apart from regular test setups

My suggestion for testing today is independent of the transaction pinning this issue is about, of course.

**Comment by elhigu** [comment]
@taybin @moll  I've considered running tests in transaction to be able to restore DB state by rolling transaction back a bad practice (like I've been considering using sqlite or h2 for testing and to use different DB in real use).

When one runs all the queries in single transaction, database cannot use multiple connections for executing queries, which makes test setup to behave quite a lot different way than real setup. For example it might hide some dead locking issues and other problems, where one has just forgot to send certain query to the same transaction by accident (that is pretty easy to do when using knex).

There are also some other subtleties how db works different inside transaction (for example how timestamps are generated etc.) in compared to multiple connections / transactions use.

Ofcourse in some specific cases running test code in singe transaction can be valid solution, but as a general practice it is not really good for testing the real use of the database.

That being said this feature would be still useful.

**Comment by moll** [comment]
> When one runs all the queries in single transaction, database cannot use multiple connections for executing queries, which makes test setup to behave quite a lot different way than real setup. For example it might hide some dead locking issues and other problems, where one has just forgot to send certain query to the same transaction by accident (that is pretty easy to do when using knex).

I've actually found that using a single connection in test (through Knex's pool config) highlights that particular problem in ways multiple connections don't. The moment one thread of execution starts a Knex transaction, all other Knex threads (`knex().transacting(tx)`), including non-transactional Knex invocations (`knex()` without `transacting`), block until that initial transaction is closed. Silently calling `BEGIN` in a before-each doesn't affect that. With multiple pooled connections you could have unassociated Knex threads proceeding in another pooled connections.

**Comment by elhigu** [comment]
@moll yeah, using pool with single connection can be useful to find those cases 👍 I've used the same technique sometimes to find accidental queries done outside of transaction, but it has also limitations. For example it also means that you cannot test code that is doing queries outside of transaction intentionally.

**Comment by mercmobily** [comment]
After 8 years, this feature request is finally being addressed! 

We're implementing this as "Connection Contexts" which will allow pinning multiple queries to the same connection without requiring a transaction: #6260

This is part of a larger effort (#6257) to implement the transaction and connection improvements originally proposed in PR #1659 by @tgriesser.

The new issue includes:
- Detailed implementation specifications
- Code examples showing the exact use case you requested
- Integration with the broader lazy transaction improvements

Thank you for your patience. Closing this in favour of the implementation issue that will deliver this functionality.

**Comment by jakepearson** [comment]
Thanks for working on this! Have a nice day.

**Comment by mercmobily** [comment]
Thank you for sticking around :D 
Realistically, this will take a little while to land. But, we will get there!!!
